### PR TITLE
Raise warning and downsample if data given to _image.resample is too large

### PR DESCRIPTION
**Pull request by dstansby**
Fixes https://github.com/matplotlib/matplotlib/issues/19276. This still needs:

- [x] Documentation
- [x] Tests

### CODE REVIEW

**Review by jklymak — CHANGES_REQUESTED** [review]
We can accurately downsample so I don't think we should error here.  An aliased downsampled image is still zoomable, at which point the aliasing will go away.  If the image is above 2**23 we should just subsample by a factor of 2 and warn.

**Comment by dstansby** [comment]
> We can accurately downsample so I don't think we should error here. An aliased downsampled image is still zoomable, at which point the aliasing will go away. If the image is above 2**23 we should just subsample by a factor of 2 and warn.

👍 - for subsampling, would you recommend just doing `arr[::2]?`

**Comment by jklymak** [comment]
I think we want `arr[::int(np.ceil(Ny/2**23)), :][:,::int(np.ceil(Nx/2**23))]`?

**Comment by dstansby** [comment]
Yep, my comment was modulo finding the right stride, but just wanted to double check that striding was the correct approach.

**Comment by jklymak** [comment]
The right approach is open to debate, in that you could anti-alias the data before hitting it with a stride.  But for such a massive subsample, I think a simple stride is better than an unspecified overflow/wrap!

**Comment by timhoffm** [comment]
Why do we bother to downsample at all? We cannot display more than N pixels and should refuse to accept more. If a user has more data, it's their responsibility and competence to reduce them.

**Comment by dstansby** [comment]
I'm happy to go either way - perhaps worth discussing on a dev call and coming back with a decision?

**Comment by jklymak** [comment]
We almost always resample so I don't see the problem with reducing the data before trying the final resample step.  This will be aliased, but a max-sized image will also be aliased by the final downsample, so I doubt it would be detectable.

**Comment by tacaswell** [comment]
I am 👍🏻 in favor of this.

**Comment by jklymak** [comment]
As discussed on the call, we should go ahead and subsample and warn, which I think is what this PR does.

**Comment by dstansby** [comment]
👍 - I still need to add a test for resampling the columns that's similar to the existing one for rows

**Comment by dstansby** [comment]
This should be good for review now. Tests and an API note added.

**Review by jklymak — APPROVED** [review]
Sorry to let this languish.  I think this looks good.

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -166,7 +167,22 @@ def _resample(
...
+    if data.shape[1] > 2**23:
+        warnings.warn(msg.format(n='2**23 rows'))
+        step = int(np.ceil(data.shape[1] / 2**23))
+        data = data[:, ::step]
+        transform = Affine2D().scale(step, 1) + transform
+    if data.shape[0] > 2**24:
+        warnings.warn(msg.format(n='2**24 columns'))
+        step = int(np.ceil(data.shape[0] / 2**24))
+        data = data[::step, :]
+        transform = Affine2D().scale(1, step) + transform
```
Comment (QuLogic) [inline lib/matplotlib/image.py:185]: 0 is the rows, 1 is the columns; you have the message swapped. I don't know if the limits should be swapped as well.

Comment (dstansby) [reply lib/matplotlib/image.py]: Thanks - I checked the limits and they are correct, just the message that needed changing.

**Comment by QuLogic** [comment]
Needs rebase for docs to work.

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
power-cycled to pick up the CI config from main.

**Pull request merged**